core/src/main/scala/org/apache/spark/sql/collection/Utils.scala

```diff
@@ -462,78 +462,84 @@ object Utils {
   def dataTypeStringBuilder(dataType: DataType,
       result: StringBuilder): Any => Unit = value => {
     dataType match {
-      case utype: UserDefinedType[_] =>
-        // check if serialized
-        if (value != null && !utype.userClass.isInstance(value)) {
-          result.append(utype.deserialize(value))
-        } else {
-          result.append(value)
-        }
-      case atype: ArrayType => value match {
+      case TimestampType => value match {
+        case l: Long => result.append(DateTimeUtils.toJavaTimestamp(l))
+        case _ => result.append(value)
+      }
+      case DateType => value match {
+        case i: Int => result.append(DateTimeUtils.toJavaDate(i))
+        case _ => result.append(value)
+      }
+      case ArrayType(elementType, _) => value match {
         case data: ArrayData =>
           result.append('[')
-          val etype = atype.elementType
           val len = data.numElements()
           if (len > 0) {
-            val ebuilder = dataTypeStringBuilder(etype, result)
-            ebuilder(data.get(0, etype))
+            val elementBuilder = dataTypeStringBuilder(elementType, result)
```

Comment:
It is not recursive in the size of the Array but the type of Array. If we have:

- Array[Long] -- level 1
- Array[Array[Long] -- level 2
- Array[Struct[Long, Int, Array[Int]]] -- level 3

So there is practically no chance of stack overflow.